### DISCUSSION TITLE
Why does conextual positining not work well on chrome/HTML ?

### DISCUSSION THREAD

**Discussion by yairc2223** [harfbuzz/harfbuzz#3469] [Q&A]
I am having trouble with the following lookups.
When I use them on Chrome, using a font loaded in an HTML, they do not work, and show this:

![10e99b74-6516-4f71-b97e-865a5d8e6245](https://user-images.githubusercontent.com/58789864/156457356-3de057e6-6d3b-442b-be4b-01dfdd8e5c35.jpg)

But they do work perfectly on VOLT proofing tool:

![5bd83470-dc73-4f7c-b9d3-97ff9a59851a](https://user-images.githubusercontent.com/58789864/156457400-f84d8e73-612e-4e2b-8c79-fd37c0f424b5.jpg)

 
If I remove the context group they work great so I am having trouble with making the context work.

Here are a minimal example of lookups:
```python
DEF_FEATURE NAME "Mark Positioning" TAG "mark"
LOOKUP "g1m"
END_FEATURE
END_LANGSYS
END_SCRIPT
DEF_GROUP "g1"
ENUM RANGE "glyph3" TO "glyph10000" END_ENUM
END_GROUP
DEF_GROUP "numbers"
ENUM RANGE "glyph37819" TO "glyph38318" END_ENUM
END_GROUP
DEF_LOOKUP "g1m" PROCESS_BASE PROCESS_MARKS ALL DIRECTION LTR
IN_CONTEXT
RIGHT GROUP "numbers"
RIGHT GROUP "numbers"
END_CONTEXT
AS_POSITION
ADJUST_SINGLE GROUP "g1" BY POS DY -200 END_POS
END_ADJUST
END_POSITION
CMAP_FORMAT 0 3 4
CMAP_FORMAT 0 4 12
CMAP_FORMAT 1 0 0
CMAP_FORMAT 3 1 4
CMAP_FORMAT 3 10 12 END
```

**Comment by behdad** [comment]
Attach the font and the test sequence here so we can debug. Does it reproduce with hb-view / hb-shape?

**Reply by yairc2223** [reply]
Hey,
actually, it is signwriting
here is the ttf file:
https://github.com/yairc2223/swfontdev/blob/ae4659eb202f750dffbecc7ff95e1616da62a7e3/yairsw.ttf

here is the HTML file:
https://github.com/yairc2223/swfontdev/blob/ae4659eb202f750dffbecc7ff95e1616da62a7e3/index1.html

**Comment by yairc2223** [comment]
Hey,
here is the ttf file:
https://github.com/yairc2223/swfontdev/blob/ae4659eb202f750dffbecc7ff95e1616da62a7e3/yairsw.ttf

here is the HTML file:
https://github.com/yairc2223/swfontdev/blob/ae4659eb202f750dffbecc7ff95e1616da62a7e3/index1.html

**Comment by behdad** [comment]
This is what I get from hb-view:
![image](https://user-images.githubusercontent.com/142891/157744410-bd97ce5d-b2a5-4fa4-9476-f9e96e7d61c4.png)
What you get in the browser might be different for other reasons.

**Reply by yairc2223** [reply]
Great ! thanks, now I am using hb-view,
I installed it successfully but what command do you run ?
I am getting:
<img width="1070" alt="Screen Shot 2022-03-14 at 18 21 21" src="https://user-images.githubusercontent.com/58789864/158220217-e1515e10-47a9-4228-8921-3ea7ad105fe4.png">

**Comment by behdad** [comment]
Try -u with unicode numbers if copy/pasting 񀀂𝤛𝤛񀀃𝤛𝤛񀀅𝤛𝤛 doesn't work in your terminal. Eg.:
```
hb-view fontfixed.ttf 񀀂𝤛𝤛񀀃𝤛𝤛񀀅𝤛𝤛
```
or:
```
hb-view fontfixed.ttf -u U+40002,U+1D91B,U+1D91B,U+40003,U+1D91B,U+1D91B,U+40005,U+1D91B,U+1D91B
```

**Reply by yairc2223** [reply]
Hey first of all tanks for all the help and support but I still encounter a problem:
My font works well on hb-view but it doesn't work at all on HTML, Do you have any insights ?
I am attaching a zipped file containing the following files.
- python script that uses hb-view and prints a png
- the ttf font SW2D
- an image output of hb-view
- an HTML that doesn't work well with the font in spite of working with hb-view

I would really like your help,
Thanks in advance

[fontHTML.zip](https://github.com/harfbuzz/harfbuzz/files/8318943/fontHTML.zip)

**Comment by behdad** [comment]
My guess is that some Unicode data in Chrome might old. cc @drott

**Reply by yairc2223** [reply]
@drott  I would like to get your insight on the matter at hand.

**Comment by drott** [comment]
For 
```
~/dev/harfbuzz/build/util/hb-view ~/dev/sw2d/fontHTML/SW2D.ttf --unicodes 1D803,1D91B,1D935,4D261,1D8F4,1D8F5,46131,1D90C,1D906,48823,1D909,1D91A --margin 1000 > sw2d.png
```
I get

<img width="514" alt="image" src="https://user-images.githubusercontent.com/1021784/160387946-e56ec3eb-7e45-4c79-b0b2-2b1d612dd065.png">

I am not sure what the expected output is, but this doesn't look right in any way, does it? 

If you need to use huge `--margin` values in `hb-view` (as in your python script in the example zip), then this indicates that spacing is still incorrect. I am sorry, I don't have time to debug this work in progress font further.

**Reply by AmitMY** [reply]
Hi @drott - this image is nearly correct.
I think the problem that Yair is facing is that in Chrome, the glyphs just stack on each other, and not actually move according to the DX/DY parameters. 

From his HTML reproduction, it looks like - 
![image](https://user-images.githubusercontent.com/5757359/160395132-a8340eb3-c02e-48eb-95b9-df044310ff5a.png)

And so the reason it is WIP is that it is unclear how to make the glyphs move in Chrome
